### PR TITLE
873 - automation of separate slick models per chain

### LINKED ISSUE #873: The source generator for slick models of database tables is not working
Currently we have multiple table namespaces that keeps separate blockchain definitions.

This isn't supported by the current app to generate the tables model.

We want to automatically generate separate files for each namespace, and put them in different scala sub-packages

### PR DESCRIPTION
**Pull request by ivanopagano**
Fixes #873

### CODE REVIEW

File: conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala [hunk 1]
```diff
@@ -1,51 +1,148 @@
...
+            fibers.traverse(_.join)
+          }.void
+
+        /* sequence the operations, you might notice we use a custom context from cats.effect
+         * to run the blocking filesystem operations on
+         */
+        for {
+          model <- getModel
+          schemas <- getSchemas(model)
+          // fiber <- blockingPool.blockOn(writeSources(schemas.toList).start)
```
Comment (olownia) [inline conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala:139]: is this commented line needed?

File: conseil-schema/src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  }
+  db {
+    driver: "org.postgresql.Driver"
+    url: "jdbc:postgresql:conseil-local"
+    user: "conseiluser"
+    password: "p@ssw0rd"
+    numThreads: 10
+    maxConnections: 10
+  }
+}
```
Comment (olownia) [inline conseil-schema/src/main/resources/application.conf:16]: no newline

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/tezos/Tables.scala [hunk 3]
```diff
@@ -1736,8 +1736,8 @@ trait Tables {
     /** Index over (level) (database name endorsing_rights_level_idx) */
     val index2 = index("endorsing_rights_level_idx", level)
 
-    /** Index over (delegate) (database name ix_delegate_slot) */
-    val index3 = index("ix_delegate_slot", delegate)
+    /** Index over (delegate,slot) (database name ix_delegate_slot) */
+    val index3 = index("ix_delegate_slot", (delegate, slot))
```
Comment (olownia) [inline conseil-common/src/main/scala/tech/cryptonomic/conseil/common/tezos/Tables.scala:1740]: is this related to this PR?

File: conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala (hunk 1 above)
Comment (ivanopagano) [reply conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala:139]: boom, thanks

File: conseil-common/src/main/scala/tech/cryptonomic/conseil/common/tezos/Tables.scala (hunk 3 above)
Comment (ivanopagano) [reply conseil-common/src/main/scala/tech/cryptonomic/conseil/common/tezos/Tables.scala:1740]: I regenerated the files with the new generator and this misalignment between definition and sources came out, so yet and no.
But I don't think it's worth its own fix PR

File: conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala [hunk 4]
```diff
@@ -1,51 +1,147 @@
...
+    val externallyProvided = args.headOption.map(
+      confFile => Paths.get(confFile).toUri.resolve(configSection)
+    )
+    val classpathProvided = new URI(configSection) //this looks up in the standard application.conf as a resource
+    println(s"""
+    | Loading schema definitions from db configurations located at
+    | - externally provided: $externallyProvided
+    | - classpath fallback: application.conf$classpathProvided
+    |
+    | The source files will be generated under the folder /tmp/conseil-slick
```
Comment (ajuszczak) [inline conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala:37]: I guess under the folder: `slickgen.codegen.outputDir`, meaning configuration file (that might create a bit confusion, that this path is hardcoded somewhere and I hope it is not :D )

File: conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala [hunk 5]
```diff
@@ -1,51 +1,147 @@
...
+    */
+  def io(
+      uri: URI,
+      outputDir: Option[String]
+  ): IO[Unit] = {
+    implicit val shift = IO.contextShift(ExecutionContext.global)
+
+    //we read configuration values
+    val dc = DatabaseConfig.forURI[PostgresProfile](uri)
+    val basePackage = dc.config.getString("codegen.package")
```
Comment (ajuszczak) [inline conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala:83]: Do we want to check this value against nullability? I believe, later in the code we may look for schema from `null.$schema`. I think it's worth to add some `println` here at least, for better user-experience. I don't think that this configuration will be missing in our case, but yet, someone may forgot or remove it for some reason, so please consider.

File: conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala [hunk 6]
```diff
@@ -1,51 +1,147 @@
...
+        val getModel = IO.fromFuture(
+          IO.delay(
+            db.run(dc.profile.createModel(None)(ExecutionContext.global).withPinnedSession)
+          )
+        )
+
+        val getSchemas = (model: Model) =>
+          IO {
+            val schemas = model.tables.groupBy(_.name.schema).collect {
+              case (Some(schema), tables) if schema != "public" => (schema, tables)
```
Comment (ajuszczak) [inline conseil-schema/src/main/scala/tech/cryptonomic/conseil/schema/GenSchema.scala:111]: I think that "public" is not the only schema visible there. We still have `information_schema`, `pg_catalog` as a default schemas, or that totally different think and I'm confusing something?

**Pull request merged**